### PR TITLE
[Feature] Support Lowcardinality optimization in LakeTable

### PR DESCRIPTION
**Pull request by abc982627271**
## What type of PR is this：
- [ ] BugFix
- [x] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 4 / 20 (20.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/statistic/StatisticExecutor.java|0|2|00.00%|[122, 123]|
|:large_blue_circle:|com/starrocks/transaction/LakeTableTxnLogApplier.java|0|13|00.00%|[34, 35, 36, 53, 54, 55, 56, 58, 59, 61, 62, 63, 64]|
|:large_blue_circle:|com/starrocks/planner/MetaScanNode.java|2|3|66.67%|[128]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|1|100.00%|[]|

File: be/src/exec/exec_node.cpp [hunk 1]
```diff
@@ -219,11 +222,15 @@ Status ExecNode::prepare(RuntimeState* state) {
 }
 
 Status ExecNode::open(RuntimeState* state) {
+    // for debug
```
Comment (liuyehcf) [inline be/src/exec/exec_node.cpp:225]: Better to remove all these debug log?

Comment (silverbullet233) [reply be/src/exec/exec_node.cpp:225]: you can use VLOG for debug

File: be/src/exec/vectorized/lake_meta_scanner.h [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+#include "storage/lake_meta_reader.h"
+#include "exec/vectorized/meta_scanner.h"
+
+
+namespace starrocks {
+namespace vectorized {
+
+class LakeMetaScanNode;
+
+class LakeMetaScanner : public MetaScanner{
```
Comment (silverbullet233) [inline be/src/exec/vectorized/lake_meta_scanner.h:19]:
```suggestion
class LakeMetaScanner final : public MetaScanner{
```

File: be/src/runtime/lake_tablets_channel.cpp [hunk 3]
```diff
@@ -137,12 +137,18 @@ class LakeTabletsChannel : public TabletsChannel {
...
 };
 
 LakeTabletsChannel::LakeTabletsChannel(LoadChannel* load_channel, const TabletsChannelKey& key, MemTracker* mem_tracker)
-        : TabletsChannel(), _load_channel(load_channel), _key(key), _mem_tracker(mem_tracker) {}
+        : TabletsChannel(), _load_channel(load_channel), _key(key), _mem_tracker(mem_tracker), 
+        _mem_pool(std::make_unique<MemPool>()) {}
 
-LakeTabletsChannel::~LakeTabletsChannel() = default;
+LakeTabletsChannel::~LakeTabletsChannel() {
+    _mem_pool.reset();
```
Comment (silverbullet233) [inline be/src/runtime/lake_tablets_channel.cpp:150]: It seems that _global_dicts will hold some pointers to the objects allocated by _mem_pool, if we destruct mem_pool first, will _global_dicts have the problem of heap-use-after-free?

File: be/src/storage/olap_meta_reader.h [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    OlapMetaReaderParams(){};
+    TabletSharedPtr tablet;
+    void check_validation() const {
+        if (UNLIKELY(version.first == -1)) {
+            LOG(FATAL) << "version is not set. tablet=" << tablet->full_name();
+        }
+    }
+};
+
+class OlapMetaReader : public MetaReader {
```
Comment (silverbullet233) [inline be/src/storage/olap_meta_reader.h:30]:
```suggestion
class OlapMetaReader final : public MetaReader {
```

File: be/src/storage/lake_meta_reader.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+struct LakeMetaReaderParams : MetaReaderParams {
+    LakeMetaReaderParams(){};
+    StatusOr<lake::Tablet> tablet;
+    std::shared_ptr<const TabletSchema> tablet_schema;
+};
+
+// MetaReader will implements
+// 1. read meta info from segment footer
+// 2. read dict info from dict page if column is dict encoding type
+class LakeMetaReader: public MetaReader {
```
Comment (silverbullet233) [inline be/src/storage/lake_meta_reader.h:28]:
```suggestion
class LakeMetaReader final : public MetaReader {
```

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 2 / 18 (11.11%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/statistic/StatisticExecutor.java|0|2|00.00%|[122, 123]|
|:large_blue_circle:|com/starrocks/transaction/LakeTableTxnLogApplier.java|0|13|00.00%|[34, 35, 36, 53, 54, 55, 56, 58, 59, 61, 62, 63, 64]|
|:large_blue_circle:|com/starrocks/planner/MetaScanNode.java|2|3|66.67%|[128]|

File: be/src/exec/exec_node.cpp (hunk 1 above)
Comment (abc982627271) [reply be/src/exec/exec_node.cpp]: removed

**Comment by dulong41** [comment]
run starrocks_admit_test

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 2 / 18 (11.11%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/statistic/StatisticExecutor.java|0|2|00.00%|[122, 123]|
|:large_blue_circle:|com/starrocks/transaction/LakeTableTxnLogApplier.java|0|13|00.00%|[34, 35, 36, 53, 54, 55, 56, 58, 59, 61, 62, 63, 64]|
|:large_blue_circle:|com/starrocks/planner/MetaScanNode.java|2|3|66.67%|[128]|

File: be/src/exec/pipeline/scan/meta_scan_context.h [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+        int32_t idx = _shared_morsel_queue ? 0 : driver_sequence;
+        if (_contexts[idx] == nullptr) {
+            _contexts[idx] = std::make_shared<MetaScanContext>(_chunk_buffer);
+        }
+        return _contexts[idx];
+    }
+
+private:
+    vectorized::MetaScanNode* const _meta_scan_node;
```
Comment (stdpain) [inline be/src/exec/pipeline/scan/meta_scan_context.h:74]: it seems unused variables

File: be/src/exec/vectorized/lake_meta_scan_node.h [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace starrocks {
+namespace vectorized {
+
+class LakeMetaScanNode final : public MetaScanNode {
+public:
+    LakeMetaScanNode(ObjectPool* pool, const TPlanNode& tnode, const DescriptorTbl& descs);
+    ~LakeMetaScanNode() override = default;
+
+    Status open(RuntimeState* state) override;
```
Comment (stdpain) [inline be/src/exec/vectorized/lake_meta_scan_node.h:18]: It seems we don't have to implements prepare/open/get_next method for LakeMetaScanNode

File: be/src/exec/vectorized/lake_meta_scanner.cpp [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+void LakeMetaScanner::close(RuntimeState* state) {
+    if (_is_closed) {
+        return;
+    }
+    if (_reader) {
+        _reader.reset();
+    }
+    _reader.reset();
```
Comment (stdpain) [inline be/src/exec/vectorized/lake_meta_scanner.cpp:68]:
```suggestion
    _reader.reset();
```

File: be/src/storage/lake_meta_reader.cpp [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+#include "runtime/global_dict/config.h"
+#include "storage/rowset/column_iterator.h"
+#include "storage/rowset/column_reader.h"
+#include "storage/rowset/rowset.h"
+
+namespace starrocks::vectorized {
+
+LakeMetaReader::LakeMetaReader() : MetaReader() {}
+
+LakeMetaReader::~LakeMetaReader() {
```
Comment (stdpain) [inline be/src/storage/lake_meta_reader.cpp:22]: prefer LakeMetaReader = default;

File: be/src/storage/lake_meta_reader.cpp [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+    _version = read_params.version;
+    _chunk_size = read_params.chunk_size;
+    _params = read_params;
+
+    return Status::OK();
+}
+
+Status LakeMetaReader::_build_collect_context(const LakeMetaReaderParams& read_params) {
+    _collect_context.seg_collecter_params.max_cid = 0;
+    for (auto it : *(read_params.id_to_names)) {
```
Comment (stdpain) [inline be/src/storage/lake_meta_reader.cpp:54]: we don't have to copy id->name pairs

File: be/src/storage/meta_reader.h [hunk 11]
```diff
@@ -58,13 +52,9 @@ class MetaReader {
     MetaReader();
     ~MetaReader();
```
Comment (stdpain) [inline be/src/storage/meta_reader.h:53]: be careful memory leak here
need add a virtual keyworks here

File: be/src/storage/olap_meta_reader.cpp [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+
+    _collect_context.cursor_idx = 0;
+    _is_init = true;
+    _has_more = true;
+    return Status::OK();
+}
+
+Status OlapMetaReader::_build_collect_context(const OlapMetaReaderParams& read_params) {
+    _collect_context.seg_collecter_params.max_cid = 0;
+    for (auto it : *(read_params.id_to_names)) {
```
Comment (stdpain) [inline be/src/storage/olap_meta_reader.cpp:54]: extra memory copy here

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 2 / 18 (11.11%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/statistic/StatisticExecutor.java|0|2|00.00%|[122, 123]|
|:large_blue_circle:|com/starrocks/transaction/LakeTableTxnLogApplier.java|0|13|00.00%|[34, 35, 36, 53, 54, 55, 56, 58, 59, 61, 62, 63, 64]|
|:large_blue_circle:|com/starrocks/planner/MetaScanNode.java|2|3|66.67%|[128]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 2 / 18 (11.11%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/statistic/StatisticExecutor.java|0|2|00.00%|[122, 123]|
|:large_blue_circle:|com/starrocks/transaction/LakeTableTxnLogApplier.java|0|13|00.00%|[34, 35, 36, 53, 54, 55, 56, 58, 59, 61, 62, 63, 64]|
|:large_blue_circle:|com/starrocks/planner/MetaScanNode.java|2|3|66.67%|[128]|

**Comment by abc982627271** [comment]
run starrocks_admit_test

**Comment by stdpain** [comment]
```patch
diff --git a/be/src/exec/pipeline/scan/olap_scan_context.cpp b/be/src/exec/pipeline/scan/olap_scan_context.cpp
diff --git a/be/src/exec/pipeline/scan/olap_scan_context.cpp b/be/src/exec/pipeline/scan/olap_scan_context.cpp
index b864749184..7b482e9828 100644
--- a/be/src/exec/pipeline/scan/olap_scan_context.cpp
+++ b/be/src/exec/pipeline/scan/olap_scan_context.cpp
@@ -85,7 +85,7 @@ Status OlapScanContext::parse_conjuncts(RuntimeState* state, const std::vector<E
     _conjuncts_manager.get_not_push_down_conjuncts(&_not_push_down_conjuncts);
 
     _dict_optimize_parser.set_mutable_dict_maps(state, state->mutable_query_global_dict_map());
-    _dict_optimize_parser.rewrite_conjuncts(&_not_push_down_conjuncts, state);
+    RETURN_IF_ERROR(_dict_optimize_parser.rewrite_conjuncts(&_not_push_down_conjuncts, state));
 
```
we also need do some change for rewrite_conjuncts

**Review by stdpain — DISMISSED** [review]

**Review by silverbullet233 — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Comment by abc982627271** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 2 / 16 (12.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/statistic/StatisticExecutor.java|0|2|00.00%|[122, 123]|
|:large_blue_circle:|com/starrocks/transaction/LakeTableTxnLogApplier.java|0|11|00.00%|[35, 36, 52, 55, 56, 57, 58, 60, 61, 63, 64]|
|:large_blue_circle:|com/starrocks/planner/MetaScanNode.java|2|3|66.67%|[128]|

**Review by sduzh — DISMISSED** [review]

**Comment by abc982627271** [comment]
run starrocks_admit_test

**Comment by abc982627271** [comment]
run starrocks_admit_test

**Comment by abc982627271** [comment]
run starrocks_admit_test

**Comment by abc982627271** [comment]
run starrocks_be_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 6 / 22 (27.27%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/statistic/StatisticExecutor.java|0|2|00.00%|[122, 123]|
|:large_blue_circle:|com/starrocks/transaction/LakeTableTxnLogApplier.java|0|13|00.00%|[35, 36, 37, 55, 56, 57, 58, 60, 61, 63, 64, 65, 66]|
|:large_blue_circle:|com/starrocks/planner/MetaScanNode.java|2|3|66.67%|[128]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 6 / 22 (27.27%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/statistic/StatisticExecutor.java|0|2|00.00%|[122, 123]|
|:large_blue_circle:|com/starrocks/transaction/LakeTableTxnLogApplier.java|0|13|00.00%|[35, 36, 37, 55, 56, 57, 58, 60, 61, 63, 64, 65, 66]|
|:large_blue_circle:|com/starrocks/planner/MetaScanNode.java|2|3|66.67%|[128]|
|:large_blue_circle:|com/starrocks/planner/OlapScanNode.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/plan/PlanFragmentBuilder.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/qe/Coordinator.java|1|1|100.00%|[]|

**Review by stdpain — DISMISSED** [review]

File: be/src/exec/exec_node.cpp [hunk 13]
```diff
@@ -219,11 +222,15 @@ Status ExecNode::prepare(RuntimeState* state) {
...
+    LOG(INFO) << "enter ExecNode::open";
```
Comment (stdpain) [inline be/src/exec/exec_node.cpp:226]: why we recover debug log?

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 2 / 18 (11.11%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/statistic/StatisticExecutor.java|0|2|00.00%|[122, 123]|
|:large_blue_circle:|com/starrocks/transaction/LakeTableTxnLogApplier.java|0|13|00.00%|[35, 36, 37, 55, 56, 57, 58, 60, 61, 63, 64, 65, 66]|
|:large_blue_circle:|com/starrocks/planner/MetaScanNode.java|2|3|66.67%|[128]|

**Comment by abc982627271** [comment]
run starrocks_be_unittest

**Review by silverbullet233 — APPROVED** [review]

**Review by stdpain — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Review by sduzh — APPROVED** [review]

**Pull request merged**